src/test/java/com/google/devtools/build/lib/runtime/BuildEventStreamerTest.java

```diff
@@ -1094,6 +1099,77 @@ public void testChunkedFlush() {
     verify(outErr, times(1)).getErr();
   }
 
+  @Test
+  public void testtestFlushPreservesStdoutStderrOrder(
+      @TestParameter({"5", "30", "10000"}) int maxBufferedLength,
+      @TestParameter({"5", "30", "10000"}) int maxChunkSize)
+      throws IOException {
+    var stdout =
+        new SynchronizedOutputStream(maxBufferedLength, maxChunkSize, /* isStderr= */ false);
+    var stderr =
+        new SynchronizedOutputStream(maxBufferedLength, maxChunkSize, /* isStderr= */ true);
+    var outErr =
+        new BuildEventStreamer.OutErrProvider() {
+          @Override
+          public Iterable<String> getOut() {
+            return stdout.readAndReset();
+          }
+
+          @Override
+          public Iterable<String> getErr() {
+            return stderr.readAndReset();
+          }
+        };
+    streamer.registerOutErrProvider(outErr);
+    stdout.registerStreamer(streamer);
+    stderr.registerStreamer(streamer);
+
+    var startEvent =
+        new GenericBuildEvent(
+            testId("Initial"), ImmutableSet.of(ProgressEvent.INITIAL_PROGRESS_UPDATE));
+    streamer.buildEvent(startEvent);
+
+    stderr.write("[0 / 3] 3 actions running\n".getBytes(UTF_8));
```

Comment:
This is a great test!

Is there any possible way to write a test where the "reconstructedOutput" ends up in a different order than the order in which the `#write()` calls happen? Or is this order preservation so effective that that just isn't possible thanks to your change?

Reply:
I don't think this scenario is possible: Whenever a state transition between stderr/stdout happens that would previously result in out-of-order reporting due to buffering, the new logic would enforce a flush. 

There is of course still a chance for discrepancies between when a `write` call is made and when it wins the CAS race (hence "best-effort"), but that doesn't seem feasible to cover in a test.